ghcide/src/Development/IDE/Plugin/CodeAction.hs

```diff
@@ -162,6 +162,61 @@ findSigOfDecl pred decls =
         any (pred . unLoc) idsSig
     ]
 
+findSigOfDeclRanged :: Range -> [LHsDecl p] -> Either String (Sig p)
+findSigOfDeclRanged range decls = do
+  dec <- findDeclContainingLocE (_start range) decls
+  case dec of
+     L _ (SigD _ sig@TypeSig {}) -> Right sig
+     L _ (ValD _ (bind :: HsBind p)) -> findSigOfBind range bind
+     _ -> Left "findSigOfDeclRanged"
+
+findSigOfBind :: Range -> HsBind p -> Either String (Sig p)
+findSigOfBind range bind =
+    case bind of
+      FunBind {} -> findSigOfLMatch (unLoc $ mg_alts (fun_matches bind))
+      _ -> Left "findSigOfBind"
+  where
+    findSigOfLMatch :: [LMatch p (LHsExpr p)] -> Either String (Sig p)
+    findSigOfLMatch ls = do
+      match <- findDeclContainingLocE (_start range) ls
+      findSigOfGRHSs (m_grhss (unLoc match))
+
+    findSigOfGRHSs :: GRHSs p (LHsExpr p) -> Either String (Sig p)
+    findSigOfGRHSs grhs = do
+        if _start range `isInsideSrcSpan` (getLoc $ grhssLocalBinds grhs)
+        then findSigOfBinds range (unLoc (grhssLocalBinds grhs)) -- where clause
+        else do
+          grhs <- findDeclContainingLocE (_start range) (grhssGRHSs grhs)
+          case unLoc grhs of
+            GRHS _ _ bd -> findSigOfExpr (unLoc bd)
+            _ -> Left "findSigOfGRHSs"
+
+    findSigOfExpr :: HsExpr p -> Either String (Sig p)
+    findSigOfExpr = go
+      where
+        go (HsLet _ binds _) = findSigOfBinds range (unLoc binds)
+        go (HsDo _ _ stmts) = do
+          stmtlr <- unLoc <$> findDeclContainingLocE (_start range) (unLoc stmts)
+          case stmtlr of
+            LetStmt _ lhsLocalBindsLR -> findSigOfBinds range $ unLoc lhsLocalBindsLR
+            _ -> Left "HsDo"
+        go _ = Left "findSigOfExpr"
+
+findSigOfBinds :: Range -> HsLocalBinds p -> Either String (Sig p)
+findSigOfBinds range = go
+  where
+    go (HsValBinds _ (ValBinds _ binds lsigs)) =
+        case unLoc <$> findDeclContainingLocE (_start range) lsigs of
+          Right sig' -> Right sig'
+          Left _ -> do
+            lHsBindLR <- findDeclContainingLocE (_start range) (bagToList binds)
+            findSigOfBind range (unLoc lHsBindLR)
+    go _ = Left "findSigOfBinds"
+
+findDeclContainingLocE :: Position -> [Located a] -> Either String (Located a)
+findDeclContainingLocE loc ls =
+  maybe (Left "findDeclContainingLoc") Right $ findDeclContainingLoc loc ls
```

Comment:
Why not use `Maybe`? Those strings in `Left` are unused, right?

Reply:
I've used them for debugging. There are too many places where a signature can appear and my `findSig`s functions are not complete. These messages have helped me locate the dead ends and extend to different kind of ast nodes. If we want to catch more cases in the future these messages can be very helpful I think.

Reply:
Well, I don't think it would be great to include these debug stuffs. Perhaps you could add some comments instead.

Reply:
I guess you're right. I'll take a look at one last thing: signatures in multiple lines and then I'll turn these `Either`\`s into `Maybe`\`s

Reply:
I've re-requested the review to wait for this change

Reply:
@kderme have you got the opprtunity of removing Eithers?

Reply:
Sorry this tool so long to address :flushed:

Reply:
No worries, thanks for taking care of requests
@berberman would it ready to go?